torchao/csrc/cuda/unpack_int4/unpack_int4.cu

```diff
@@ -0,0 +1,268 @@
+        break; \
+      case 8: \
+        _dequantize_int4_kernel<c10::BFloat16, 8, QGROUPSIZE, true><<<grid, kWarpSize, 0, stream>>>( \
+        packed_w.packed_accessor32<int32_t, 4, at::RestrictPtrTraits>(), \
+        out.packed_accessor32<c10::BFloat16, 2, at::RestrictPtrTraits>(), \
+        scales_and_zeros.packed_accessor32<c10::BFloat16, 3, at::RestrictPtrTraits>()); \
+        break; \
+      default: \
+        break; \
+    } \
+  } while(false)
+
+#define DISPATCH_Q_GROUP() \
+  do { \
+    switch (group_size) { \
+      case 32: \
+        RUN_DEQUANT(32); \
+        break; \
+      case 64: \
+        RUN_DEQUANT(64); \
+        break; \
+      case 128: \
+        RUN_DEQUANT(128); \
+        break; \
+      case 256: \
+        RUN_DEQUANT(256); \
+        break; \
+      default: \
+        break; \
+    } \
+  } while(false)
+
+  DISPATCH_Q_GROUP();
+  #undef DISPATCH_Q_GROUP
+  #undef RUN_DEQUANT
+
+  return out;
+}
+
+// output is [n][k] (int32 dtype)
+// input is [n / 8][k / (InnerKTiles * 16)][32][innerKTiles / 2]
+at::Tensor _unpack_int4_to_int(
+    const at::Tensor& packed_w,
+    int64_t innerKTiles) 
+{
+
+  c10::cuda::CUDAGuard g(packed_w.device());
+
+  TORCH_CHECK(packed_w.dim() == 4);
+  TORCH_CHECK(packed_w.dtype() == at::kInt);
+  TORCH_CHECK(packed_w.is_contiguous());
+
+  TORCH_CHECK(packed_w.size(2) == 32);
+  TORCH_CHECK(packed_w.size(3) == innerKTiles / 2);
+  TORCH_CHECK(innerKTiles == 2 || innerKTiles == 4 || innerKTiles == 8);
+
+  int N = packed_w.size(0) * 8;
+  int K = packed_w.size(1) * innerKTiles * 16;
+
+  constexpr int32_t kNTileSize = 8;
+  constexpr int32_t kKTileSize = 16;
+
+  auto nTiles = divUp(N, kNTileSize);
+
+  auto kSuperTiles = divUp(K, innerKTiles * kKTileSize);
+
+  auto out = at::empty(
+      {N, K},
+      at::TensorOptions().dtype(at::kInt).device(packed_w.device()));
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+  dim3 grid(kSuperTiles, nTiles);
+  
+  if (innerKTiles == 2) {
+    _dequantize_int4_kernel<int32_t, 2, 0, false><<<grid, kWarpSize, 0, stream>>>(
+        packed_w.packed_accessor32<int32_t, 4, at::RestrictPtrTraits>(),
+        out.packed_accessor32<int32_t, 2, at::RestrictPtrTraits>());
+  }
+   else if (innerKTiles == 4) {
+    _dequantize_int4_kernel<int32_t, 4, 0, false><<<grid, kWarpSize, 0, stream>>>(
+        packed_w.packed_accessor32<int32_t, 4, at::RestrictPtrTraits>(),
+        out.packed_accessor32<int32_t, 2, at::RestrictPtrTraits>());
+  } else if (innerKTiles == 8) {
+    _dequantize_int4_kernel<int32_t, 8, 0, false><<<grid, kWarpSize, 0, stream>>>(
+        packed_w.packed_accessor32<int32_t, 4, at::RestrictPtrTraits>(),
+        out.packed_accessor32<int32_t, 2, at::RestrictPtrTraits>());
+  }
+
+  return out;
+}
+
+TORCH_LIBRARY_IMPL(torchao, CUDA, m) {
+  m.impl("torchao::unpack_int4_to_int", &_unpack_int4_to_int);
+  m.impl("torchao::dequantize_int4", &_dequantize_int4);
```

Comment:
nit: I feel we probably need to mention `tensor_core_tiled` layout in the name of these ops if these are specific to that packing format